kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/stores/StreamThreadFlushListeners.java

```diff
@@ -0,0 +1,199 @@
+ * AsyncProcessor instance that it needs to flush. Unfortunately, this is
+ * significantly harder than it would seem, due to the order of operations
+ * and slow drip of information such as which StreamTask/partition the store
+ * and processor correspond to.
+ * The difficulty is this:
+ * Even though the ProcessorSupplier has a handle on all the AsyncStoreBuilders,
+ * the actual state store instance itself is not yet created when the processor
+ * supplier constructs a new processor, so we can't connect the new AsyncProcessor
+ * instance to any stores at this time. And even though the AsyncProcessor we
+ * want to connect to does exist when the StoreBuilder goes to build the
+ * AsyncFlushingStore layer for a newly constructed state store, we have no way
+ * of getting a handle on the specific AsyncProcessor that was just created from
+ * within the #build method. Even if the AsyncProcessor passed itself to the
+ * StoreBuilder and the builder saved a reference to each new processor, we wouldn't
+ * know which processor to connect to when #build is called, since the StoreBuilder
+ * is used by all StreamThreads across all their StreamTasks. We can only resolve
+ * the StreamThread at this point, but not the StreamTask, since neither the
+ * AsyncProcessor nor the AsyncFlushingStore have any way to tell which
+ * task/partition they are associated with from within their constructors.
+ * So, we need to wait until both the processor and the store are initialized,
+ * and Streams passes the ProcessorContext into #init from which we can at last
+ * extract the partition.
+ * Putting it all together, since the processor is initialized first, it will
+ * register a callback that triggers flush on itself, along with the partition
+ * that flush callback (ie {@link AsyncFlushListener}) is mapped to, with the
+ * StoreBuilder. The StoreBuilder itself maintains a map from each StreamThread
+ * to the metadata for this store/StoreBuilder, including a map from partition
+ * to AsyncFlushListener that it builds up as processors are initialized.
+ * Then when the state store is initialized at long last and gets the context
+ * that includes the partition information, it can use its stored reference
+ * to the metadata object from the StoreBuilder to finally retrieve the
+ * AsyncFlushListener that is associated with the same partition.
+ * <p>
+ * TODO: This is obviously super convoluted so we should investigate whether
+ *  there is a better way to connect up the flushing layer with the processor
+ *  it needs to flush. This may not be possible without extensive changes to
+ *  Kafka Streams itself, but it's worth exploring.
+ * <p>
+ * Threading notes:
+ * -only ever accessed by/on StreamThreads
+ * -each StoreBuilder creates exactly one of these for each StreamThread
+ * -should only ever be accessed by that particular StreamThread, so no
+ *  concurrency precautions are taken within this class
+ */
+public class StreamThreadFlushListeners {
```

Comment:
If we don't go the route of using a designated no-op store, I think a simpler way of injecting these would be to use the `ResponsiveStoreRegistry`. I'm working on a PR that has a similar use case for injecting the processor context into `PartitionedOperations`, which we need to do safe concurrent store access. You can build off that.

Reply:
https://github.com/responsivedev/responsive-pub/pull/270

Reply:
This is definitely a mess of a class. If I can't figure out how to make your no-op designated flushing store idea work, I'll take a look into simplifying this using the registry (though probably in the followup PR with things like the punctuator-based flushing)